### ISSUE TITLE
Rule proposal: `no-restricted-loops`

### ISSUE THREAD

**Issue by zanminkian** [sindresorhus/eslint-plugin-unicorn#2453]
### Description

JavaScript has 6 kind of loops. We don't really need all of them. Just allow 2 of them (by default) is enough.

  |Loop|Example|Allow(by default)|
  |----|----|----|
  |while|`while(true){}`|✅|
  |for-of|`for(const foo of bar){}`|✅|
  |for|`for(let i=0;i<10;i++){}`|❌|
  |for-in|`for(const foo in bar){}`|❌|
  |do-while|`do{}while(foo<bar)`|❌|
  |for-await-of|`for await (const num of foo()){}`|❌|

This rule provide an options: `restricted-loops`, defaults to `['for','for-in','do-while','for-await-of']`

### Fail

See Description.

### Pass

See Description.

### Proposed rule name

no-restricted-loops

**Comment by fregante** [comment]
Each loop serves a different purpose. I think only `for` and `for-in` should generally be avoided because of their possible issues.

https://typescript-eslint.io/rules/prefer-for-of/
https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-for-loop.md

I'm surprised that there's no rule or rule suggestion for `for-in` though. @fisker did I miss any?

**Comment by fisker** [comment]
> I'm surprised that there's no rule or rule suggestion for for-in though.

There seems no rule forbids `for..in` loop, but only [`guard-for-in`](https://eslint.org/docs/latest/rules/guard-for-in).


Fun fact:  This is valid 😄 

```js
for (var property = 1 in object);
//                ^^^
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Deprecated_and_obsolete_features#statements

**Comment by sindresorhus** [comment]
We are passing on this.
